src/utils/RouteUtils.js

```diff
@@ -1,8 +1,162 @@
+import AllStopUtils from './AllStopUtils';
+import GraphhopperUtils from './GraphhopperUtils';
 import LogUtils from './LogUtils';
 import ParseRouteUtils from './ParseRouteUtils';
 
+/**
+ * Returns the flattened version of arr.
+ * @param arr
+ * @returns {Array<Object>}
+ */
+function flatten(arr: Array<Array<Object>>): Array<Object> {
+  return [].concat(...arr);
+}
+
+/**
+ * Returns whether or not location is a bus stop.
+ * @param location
+ * @returns {Promise<boolean>}
+ */
+async function isBusStop(location: string): Promise<boolean> {
+  const stops = await AllStopUtils.fetchAllStops();
+  return stops.filter(s => s.name === location).length > 0;
+}
+
+/**
+ * Filter and validate the array of parsed routes to send to the client
+ * and then return routes categorized as fromStop, boardingSoon, or walking.
+ *
+ * @param routeBus
+ * @param routeWalking
+ * @param start
+ * @param end
+ * @param departureTimeQuery
+ * @param isArriveBy
+ * @param originBusStopName
+ * @returns {Promise<Object>}
+ */
+async function createSectionedRoutes(
+  busRoutes: Array<Object>,
+  walkingRoute: Object,
+  start: string,
+  end: string,
+  departureTimeQuery: number,
+  isArriveBy: boolean,
+  originBusStopName: ?string,
+): Promise<Object> {
+  const departureTimeNowMs = parseFloat(departureTimeQuery) * 1000;
+  const departureDelayBuffer = !isArriveBy;
+
+  const startPoint = ParseRouteUtils.latLongFromStr(start);
+  const endPoint = ParseRouteUtils.latLongFromStr(end);
+
+  const finalRoutes = (await Promise.all(
+    busRoutes.map(currPath => ParseRouteUtils.condenseRoute(
+      currPath,
+      startPoint,
+      endPoint,
+      walkingRoute.directions[0].distance,
+      departureDelayBuffer,
+      departureTimeNowMs,
+    )),
+  )).filter(route => route !== null);
+
+  const sectionedRoutes = {
+    boardingSoon: [],
+    fromStop: [],
+    walking: [],
+  };
+
+  if (walkingRoute) { // if a walkingRoute exists append it
+    sectionedRoutes.walking.push(walkingRoute);
+  }
+
+  finalRoutes.forEach((r) => {
+    if (originBusStopName !== null) {
+      const { directions } = r;
+      if (directions.length > 0) {
+        const { stops } = directions[0];
+        if (stops.length > 0 && stops[0].name === originBusStopName) {
+          sectionedRoutes.fromStop.push(r);
+          return;
+        }
+      }
+    }
+    sectionedRoutes.boardingSoon.push(r);
+  });
+
+  return sectionedRoutes;
+}
+
+/**
+ * Returns the routes for a search categorizing them as being fromStop,
+ * boardingSoon, or walking.
+ *
+ * @param destinationName
+ * @param end
+ * @param start
+ * @param departureTimeQuery
+ * @param isArriveBy
+ * @param originBusStopName
+ * @returns {Promise<Object>}
+ */
+async function getSectionedRoutes(
+  destinationName: string,
+  end: string,
+  start: string,
+  departureTimeQuery: number,
+  isArriveBy: boolean,
+  originBusStopName: ?string,
+): Promise<Object> {
+  const defaultState = {
```

Comment:
i don't know if `defaultState` is the best name here. maybe `defaultRoutes` or something

Reply:
yeah i made the change to keep the types consistent